README.md
# primus-socket.io-client

[![Version npm][npm-primus-socket.io-client-badge]][npm-primus-socket.io-client][![Build Status][travis-primus-socket.io-client-badge]][travis-primus-socket.io-client][![Dependencies][david-primus-socket.io-client-badge]][david-primus-socket.io-client][![Coverage Status][coverage-primus-socket.io-client-badge]][coverage-primus-socket.io-client][![IRC channel][irc-badge]][irc]

[npm-primus-socket.io-client-badge]: https://img.shields.io/npm/v/primus-socket.io-client.svg?style=flat-square
[npm-primus-socket.io-client]: http://browsenpm.org/package/primus-socket.io-client
[travis-primus-socket.io-client-badge]: https://img.shields.io/travis/primus/primus-socket.io-client/master.svg?style=flat-square
[travis-primus-socket.io-client]: https://travis-ci.org/primus/primus-socket.io-client
[david-primus-socket.io-client-badge]: https://img.shields.io/david/primus/primus-socket.io-client.svg?style=flat-square
[david-primus-socket.io-client]: https://david-dm.org/primus/primus-socket.io-client
[coverage-primus-socket.io-client-badge]: https://img.shields.io/coveralls/primus/primus-socket.io-client/master.svg?style=flat-square
[coverage-primus-socket.io-client]: https://coveralls.io/r/primus/primus-socket.io-client?branch=master
[irc-badge]: https://img.shields.io/badge/IRC-irc.freenode.net%23primus-00a8ff.svg?style=flat-square
[irc]: https://webchat.freenode.net/?channels=primus

This is a fork of the original Socket.IO client's 0.9 codebase. We've forked it
to provide additional stability fixes. One of the main reasons we still maintain
this fork is that using Socket.IO 1.0 inside of Primus doesn't make sense as
Socket.IO 1.0 is just a wrapper around Engine.IO so it makes more sense for our
users to use our `engine.io` transformer instead of supporting Socket.IO 1.0.

#### Sockets for the rest of us

The `socket.io` client is basically a simple HTTP Socket interface implementation.
It looks similar to WebSocket while providing additional features and
leveraging other transports when WebSocket is not supported by the user's
browser.

```js
var socket = io.connect('http://domain.com');
socket.on('connect', function () {
  // socket connected
});
socket.on('custom event', function () {
  // server emitted a custom event
});
socket.on('disconnect', function () {
  // socket disconnected
});
socket.send('hi there');
```

### Recipes

#### Utilizing namespaces (ie: multiple sockets)

If you want to namespace all the messages and events emitted to a particular
endpoint, simply specify it as part of the `connect` uri:

```js
var chat = io.connect('http://localhost/chat');
chat.on('connect', function () {
  // chat socket connected
});

var news = io.connect('/news'); // io.connect auto-detects host
news.on('connect', function () {
  // news socket connected
});
```

#### Emitting custom events

To ease with the creation of applications, you can emit custom events outside
of the global `message` event.

```js
var socket = io.connect();
socket.emit('server custom event', { my: 'data' });
```

#### Forcing disconnection

```js
var socket = io.connect();
socket.on('connect', function () {
  socket.disconnect();
});
```

### Documentation

#### io#connect

```js
io.connect(uri, [options]);
```

##### Options:

- *resource*

    socket.io

  The resource is what allows the `socket.io` server to identify incoming connections by `socket.io` clients. In other words, any HTTP server can implement socket.io and still serve other normal, non-realtime HTTP requests.

- *transports*

```js
['websocket', 'flashsocket', 'htmlfile', 'xhr-multipart', 'xhr-polling', 'jsonp-polling']
```

  A list of the transports to attempt to utilize (in order of preference).

- *'connect timeout'*

```js
5000
```

  The amount of milliseconds a transport has to create a connection before we consider it timed out.

- *'try multiple transports'*

```js
true
```

  A boolean indicating if we should try other transports when the  connectTimeout occurs.

- *reconnect*

```js
true
```

  A boolean indicating if we should automatically reconnect if a connection is disconnected.

- *'reconnection delay'*

```js
500
```

  The amount of milliseconds before we try to connect to the server again. We are using a exponential back off algorithm for the following reconnections, on each reconnect attempt this value will get multiplied (500 > 1000 > 2000 > 4000 > 8000).


- *'max reconnection attempts'*

```js
10
```

  The amount of attempts should we make using the current transport to connect to the server? After this we will do one final attempt, and re-try with all enabled transport methods before we give up.

##### Properties:

- *options*

  The passed in options combined with the defaults.

- *connected*

  Whether the socket is connected or not.

- *connecting*

  Whether the socket is connecting or not.

- *reconnecting*

  Whether we are reconnecting or not.

- *transport*

  The transport instance.

##### Methods:

- *connect(λ)*

  Establishes a connection. If λ is supplied as argument, it will be called once the connection is established.

- *send(message)*

  A string of data to send.

- *disconnect*

  Closes the connection.

- *on(event, λ)*

  Adds a listener for the event *event*.

- *once(event, λ)*

  Adds a one time listener for the event *event*. The listener is removed after the first time the event is fired.

- *removeListener(event, λ)*

  Removes the listener λ for the event *event*.

##### Events:

- *connect*

  Fired when the connection is established and the handshake successful.

- *connecting(transport_type)*

    Fired when a connection is attempted, passing the transport name.

- *connect_failed*

    Fired when the connection timeout occurs after the last connection attempt.
  This only fires if the `connectTimeout` option is set.
  If the `tryTransportsOnConnectTimeout` option is set, this only fires once all
  possible transports have been tried.

- *message(message)*

  Fired when a message arrives from the server

- *close*

  Fired when the connection is closed. Be careful with using this event, as some transports will fire it even under temporary, expected disconnections (such as XHR-Polling).

- *disconnect*

  Fired when the connection is considered disconnected.

- *reconnect(transport_type,reconnectionAttempts)*

  Fired when the connection has been re-established. This only fires if the `reconnect` option is set.

- *reconnecting(reconnectionDelay,reconnectionAttempts)*

  Fired when a reconnection is attempted, passing the next delay for the next reconnection.

- *reconnect_failed*

  Fired when all reconnection attempts have failed and we where unsuccessful in reconnecting to the server.

### Contributors

Guillermo Rauch &lt;guillermo@learnboost.com&gt;

Arnout Kazemier &lt;info@3rd-eden.com&gt;

### License

[MIT](LICENSE)


History.md
2.0.0 / 2016-03-05
==================

 * Remove workaround for iOS 5 WebSocket bug

1.0.3 / 2016-01-06
==================

 * Bump ws to version 1.0.x

1.0.2 / 2015-08-23
==================

 * Bump ws to version 0.8.x

1.0.1 / 2015-03-17
==================

 * Minor license update

1.0.0 / 2015-03-15
==================

 * Fix a bug in the namespacing used for browser builds
 * Abort the connection if the socket gets closed during the handshake

0.9.17 / 2014-05-22
===================

 * bump

0.9.16 / 2013-06-06
===================

  * transports: fix escaping for tests

0.9.15 / 2013-06-06
===================

  * transports: added unescaping for escaped htmlfile
  * skipped 12-14 to match socket.io server version

0.9.11 / 2012-11-02
===================

  * Enable use of 'xhr' transport in Node.js
  * Fix the problem with disconnecting xhr-polling users
  * Add should to devDependencies
  * Prefer XmlHttpRequest if CORS is available
  * Make client compatible with AMD loaders.

0.9.10 / 2012-08-10
===================

  * fix removeAllListeners to behave as expected.
  * set withCredentials to true only if xdomain.
  * socket: disable disconnect on unload by default.

0.9.9 / 2012-08-01
==================

  * socket: fixed disconnect xhr url and made it actually sync
  * *: bump xmlhttprequest dep

0.9.8 / 2012-07-24
==================

  * Fixed build.

0.9.7 / 2012-07-24
==================

  * iOS websocket crash fix.
  * Fixed potential `open` collision.
  * Fixed disconnectSync.

0.9.6 / 2012-04-17
==================

  * Don't position the jsonp form off the screen (android fix).

0.9.5 / 2012-04-05
==================

  * Bumped version.

0.9.4 / 2012-04-01
==================

  * Fixes polling loop upon reconnect advice (fixes #438).

0.9.3 / 2012-03-28
==================

  * Fix XHR.check, which was throwing an error transparently and causing non-IE browsers to fall back to JSONP [mikito]
  * Fixed forced disconnect on window close [zzzaaa]

0.9.2 / 2012-03-13
==================

  * Transport order set by "options" [zzzaaa]

0.9.1-1 / 2012-03-02
====================

  * Fixed active-x-obfuscator NPM dependency.

0.9.1 / 2012-03-02
==================

  * Misc corrections.
  * Added warning within Firefox about webworker test in test runner.
  * Update ws dependency [einaros]
  * Implemented client side heartbeat checks. [felixge]
  * Improved Firewall support with ActiveX obfuscation. [felixge]
  * Fixed error handling during connection process. [Outsideris]

0.9.0 / 2012-02-26
==================

  * Added DS_Store to gitignore.
  * Updated depedencies.
  * Bumped uglify
  * Tweaking code so it doesn't throw an exception when used inside a WebWorker in Firefox
  * Do not rely on Array.prototype.indexOf as it breaks with pages that use the Prototype.js library.
  * Windows support landed
  * Use @einaros ws module instead of the old crap one
  * Fix for broken closeTimeout and 'IE + xhr' goes into infinite loop on disconnection
  * Disabled reconnection on error if reconnect option is set to false
  * Set withCredentials to true before xhr to fix authentication
  * Clears the timeout from reconnection attempt when there is a successful or failed reconnection. 
    This fixes the issue of setTimeout's carrying over from previous reconnection
    and changing (skipping) values of self.reconnectionDelay in the newer reconnection.
  * Removed decoding of parameters when chunking the query string.
    This was used later on to construct the url to post to the socket.io server
    for connection and if we're adding custom parameters of our own to this url
    (for example for OAuth authentication) they were being sent decoded, which is wrong.

0.8.7 / 2011-11-05
==================

  * Bumped client

0.8.6 / 2011-10-27
==================

  * Added WebWorker support.
  * Fixed swfobject and web_socket.js to not assume window.
  * Fixed CORS detection for webworker.
  * Fix `defer` for webkit in a webworker.
  * Fixed io.util.request to not rely on window.
  * FIxed; use global instead of window and dont rely on document.
  * Fixed; JSON-P handshake if CORS is not available.
  * Made underlying Transport disconnection trigger immediate socket.io disconnect.
  * Fixed warning when compressing with Google Closure Compiler.
  * Fixed builder's uglify utf-8 support.
  * Added workaround for loading indicator in FF jsonp-polling. [3rd-Eden]
  * Fixed host discovery lookup. [holic]
  * Fixed close timeout when disconnected/reconnecting. [jscharlach]
  * Fixed jsonp-polling feature detection.
  * Fixed jsonp-polling client POSTing of \n.
  * Fixed test runner on IE6/7

0.8.5 / 2011-10-07
==================

  * Bumped client

0.8.4 / 2011-09-06
==================

  * Corrected build

0.8.3 / 2011-09-03
==================

  * Fixed `\n` parsing for non-JSON packets.
  * Fixed; make Socket.IO XHTML doctype compatible (fixes #460 from server)
  * Fixed support for Node.JS running `socket.io-client`.
  * Updated repository name in `package.json`.
  * Added support for different policy file ports without having to port
    forward 843 on the server side [3rd-Eden]

0.8.2 / 2011-08-29
==================

  * Fixed flashsocket detection.

0.8.1 / 2011-08-29
==================

  * Bump version.

0.8.0 / 2011-08-28
==================

  * Added MozWebSocket support (hybi-10 doesn't require API changes) [einaros].

0.7.11 / 2011-08-27
===================

  * Corrected previous release (missing build).

0.7.10 / 2011-08-27
===================

  * Fix for failing fallback in websockets

0.7.9 / 2011-08-12
==================

  * Added check on `Socket#onConnect` to prevent double `connect` events on the main manager.
  * Fixed socket namespace connect test. Remove broken alternative namespace connect test.
  * Removed test handler for removed test.
  * Bumped version to match `socket.io` server.

0.7.5 / 2011-08-08
==================


lib/socket.js
(function (exports, io, global) {

  /**
   * Expose constructor.
   */

  exports.Socket = Socket;

  /**
   * Create a new `Socket.IO client` which can establish a persistent
   * connection with a Socket.IO enabled server.
   *
   * @api public
   */

  function Socket (options) {
    this.options = {
        port: 80
      , secure: false
      , document: 'document' in global ? document : false
      , resource: 'socket.io'
      , transports: io.transports
      , 'connect timeout': 10000
      , 'try multiple transports': true
      , 'reconnect': true
      , 'reconnection delay': 500
      , 'reconnection limit': Infinity
      , 'reopen delay': 3000
      , 'max reconnection attempts': 10
      , 'sync disconnect on unload': false
      , 'auto connect': true
      , 'flash policy port': 10843
      , 'manualFlush': false
    };

    io.util.merge(this.options, options);

    this.connected = false;
    this.open = false;
    this.connecting = false;
    this.reconnecting = false;
    this.namespaces = {};
    this.buffer = [];
    this.doBuffer = false;

    if (this.options['sync disconnect on unload'] &&
        (!this.isXDomain() || io.util.ua.hasCORS)) {
      var self = this;
      io.util.on(global, 'beforeunload', function () {
        self.disconnectSync();
      }, false);
    }

    if (this.options['auto connect']) {
      this.connect();
    }
  }

  /**
   * Apply EventEmitter mixin.
   */

  io.util.mixin(Socket, io.EventEmitter);

  /**
   * Returns a namespace listener/emitter for this socket
   *
   * @api public
   */

  Socket.prototype.of = function (name) {
    if (!this.namespaces[name]) {
      this.namespaces[name] = new io.SocketNamespace(this, name);

      if (name !== '') {
        this.namespaces[name].packet({ type: 'connect' });
      }
    }

    return this.namespaces[name];
  };

  /**
   * Emits the given event to the Socket and all namespaces
   *
   * @api private
   */

  Socket.prototype.publish = function () {
    this.emit.apply(this, arguments);

    var nsp;

    for (var i in this.namespaces) {
      if (this.namespaces.hasOwnProperty(i)) {
        nsp = this.of(i);
        nsp.$emit.apply(nsp, arguments);
      }
    }
  };

  /**
   * Performs the handshake
   *
   * @api private
   */

  function empty () { };

  Socket.prototype.handshake = function (fn) {
    var self = this
      , options = this.options;

    function complete (data) {
      if (data instanceof Error) {
        self.connecting = false;
        return self.onError(data.message);
      }
      if (!self.connecting) return;

      fn.apply(null, data.split(':'));
    }

    var url = [
          'http' + (options.secure ? 's' : '') + ':/'
        , options.host + ':' + options.port
        , options.resource
        , io.protocol
        , io.util.query(this.options.query, 't=' + +new Date)
      ].join('/');

    if (this.isXDomain() && !io.util.ua.hasCORS) {
      var insertAt = document.getElementsByTagName('script')[0]
        , script = document.createElement('script');

      script.src = url + '&jsonp=' + io.j.length;
      insertAt.parentNode.insertBefore(script, insertAt);

      io.j.push(function (data) {
        complete(data);
        script.parentNode.removeChild(script);
      });
    } else {
      var xhr = io.util.request();

      xhr.open('GET', url, true);
      if (this.isXDomain()) {
        xhr.withCredentials = true;
      }
      xhr.onreadystatechange = function () {
        if (xhr.readyState == 4) {
          xhr.onreadystatechange = empty;

          if (xhr.status == 200) {
            complete(xhr.responseText);
          } else if (xhr.status == 403) {
            self.onError(xhr.responseText);
          } else {
            self.connecting = false;
            !self.reconnecting && self.onError(xhr.responseText);
          }
        }
      };
      xhr.send(null);
    }
  };

  /**
   * Find an available transport based on the options supplied in the constructor.
   *
   * @api private
   */

  Socket.prototype.getTransport = function (override) {
    var transports = override || this.transports, match;

    for (var i = 0, transport; transport = transports[i]; i++) {
      if (io.Transport[transport]
        && io.Transport[transport].check(this)
        && (!this.isXDomain() || io.Transport[transport].xdomainCheck(this))) {
        return new io.Transport[transport](this, this.sessionid);
      }
    }

    return null;
  };

  /**
   * Connects to the server.
   *
   * @param {Function} [fn] Callback.
   * @returns {io.Socket}
   * @api public
   */

  Socket.prototype.connect = function (fn) {
    if (this.connecting) {
      return this;
    }

    var self = this;
    self.connecting = true;

    this.handshake(function (sid, heartbeat, close, transports) {
      self.sessionid = sid;
      self.closeTimeout = close * 1000;
      self.heartbeatTimeout = heartbeat * 1000;
      if(!self.transports)
          self.transports = self.origTransports = (transports ? io.util.intersect(
              transports.split(',')
            , self.options.transports
          ) : self.options.transports);

      self.setHeartbeatTimeout();

      function connect (transports){
        if (self.transport) self.transport.clearTimeouts();

        self.transport = self.getTransport(transports);
        if (!self.transport) return self.publish('connect_failed');

        // once the transport is ready
        self.transport.ready(self, function () {
          self.connecting = true;
          self.publish('connecting', self.transport.name);
          self.transport.open();

          if (self.options['connect timeout']) {
            self.connectTimeoutTimer = setTimeout(function () {
              if (!self.connected) {
                self.connecting = false;

                if (self.options['try multiple transports']) {
                  var remaining = self.transports;

                  while (remaining.length > 0 && remaining.splice(0,1)[0] !=
                         self.transport.name) {}

                    if (remaining.length){
                      connect(remaining);
                    } else {
                      self.publish('connect_failed');
                    }
                }
              }
            }, self.options['connect timeout']);
          }
        });
      }

      connect(self.transports);

      self.once('connect', function (){
        clearTimeout(self.connectTimeoutTimer);

        fn && typeof fn == 'function' && fn();
      });
    });

    return this;
  };

  /**
   * Clears and sets a new heartbeat timeout using the value given by the
   * server during the handshake.
   *
   * @api private
   */

  Socket.prototype.setHeartbeatTimeout = function () {
    clearTimeout(this.heartbeatTimeoutTimer);
    if(this.transport && !this.transport.heartbeats()) return;

    var self = this;
    this.heartbeatTimeoutTimer = setTimeout(function () {
      self.transport.onClose();
    }, this.heartbeatTimeout);
  };

  /**
   * Sends a message.
   *
   * @param {Object} data packet.
   * @returns {io.Socket}
   * @api public
   */

  Socket.prototype.packet = function (data) {
    if (this.connected && !this.doBuffer) {
      this.transport.packet(data);
    } else {
      this.buffer.push(data);
    }

    return this;
  };

  /**
   * Sets buffer state
   *
   * @api private
   */

  Socket.prototype.setBuffer = function (v) {
    this.doBuffer = v;

    if (!v && this.connected && this.buffer.length) {
      if (!this.options['manualFlush']) {
        this.flushBuffer();
      }
    }
  };

  /**
   * Flushes the buffer data over the wire.
   * To be invoked manually when 'manualFlush' is set to true.
   *
   * @api public
   */

  Socket.prototype.flushBuffer = function() {
    this.transport.payload(this.buffer);
    this.buffer = [];
  };

  /**
   * Disconnect the established connect.
   *
   * @returns {io.Socket}
   * @api public
   */

  Socket.prototype.disconnect = function () {
    if (this.connected || this.connecting) {
      if (this.open) {
        this.of('').packet({ type: 'disconnect' });
      }

      // handle disconnection immediately
      this.onDisconnect('booted');
    }

    return this;
  };

  /**
   * Disconnects the socket with a sync XHR.
   *
   * @api private
   */

  Socket.prototype.disconnectSync = function () {
    // ensure disconnection
    var xhr = io.util.request();
    var uri = [
        'http' + (this.options.secure ? 's' : '') + ':/'
      , this.options.host + ':' + this.options.port
      , this.options.resource
      , io.protocol
      , ''
      , this.sessionid
    ].join('/') + '/?disconnect=1';

    xhr.open('GET', uri, false);
    xhr.send(null);

    // handle disconnection immediately
    this.onDisconnect('booted');
  };

  /**
   * Check if we need to use cross domain enabled transports. Cross domain would
   * be a different port or different domain name.
   *
   * @returns {Boolean}
   * @api private
   */

  Socket.prototype.isXDomain = function () {
    // if node
    return false;
    // end node

    var port = global.location.port ||
      ('https:' == global.location.protocol ? 443 : 80);

    return this.options.host !== global.location.hostname
      || this.options.port != port;
  };

  /**
   * Called upon handshake.
   *
   * @api private
   */

  Socket.prototype.onConnect = function () {
    if (!this.connected) {
      this.connected = true;
      this.connecting = false;
      if (!this.doBuffer) {
        // make sure to flush the buffer
        this.setBuffer(false);
      }
      this.emit('connect');
    }
  };

  /**
   * Called when the transport opens
   *
   * @api private
   */

  Socket.prototype.onOpen = function () {
    this.open = true;
  };

  /**
   * Called when the transport closes.
   *
   * @api private
   */

  Socket.prototype.onClose = function () {
    this.open = false;
    clearTimeout(this.heartbeatTimeoutTimer);
  };

  /**
   * Called when the transport first opens a connection
   *
   * @param text
   */

  Socket.prototype.onPacket = function (packet) {
    this.of(packet.endpoint).onPacket(packet);
  };

  /**
   * Handles an error.
   *
   * @api private
   */

  Socket.prototype.onError = function (err) {
    if (err && err.advice) {
      if (err.advice === 'reconnect' && (this.connected || this.connecting)) {
        this.disconnect();
        if (this.options.reconnect) {
          this.reconnect();
        }
      }
    }

    this.publish('error', err && err.reason ? err.reason : err);
  };

  /**
   * Called when the transport disconnects.
   *
   * @api private
   */

  Socket.prototype.onDisconnect = function (reason) {
    var wasConnected = this.connected
      , wasConnecting = this.connecting;

    this.connected = false;
    this.connecting = false;
    this.open = false;

    if (wasConnected || wasConnecting) {
      this.transport.close();
      this.transport.clearTimeouts();
      if (wasConnected) {
        this.publish('disconnect', reason);

        if ('booted' != reason && this.options.reconnect && !this.reconnecting) {
          this.reconnect();
        }
      }
    }
  };

  /**
   * Called upon reconnection.
   *
   * @api private
   */

  Socket.prototype.reconnect = function () {
    this.reconnecting = true;
    this.reconnectionAttempts = 0;
    this.reconnectionDelay = this.options['reconnection delay'];

    var self = this
      , maxAttempts = this.options['max reconnection attempts']
      , tryMultiple = this.options['try multiple transports']
      , limit = this.options['reconnection limit'];

    function reset () {
      if (self.connected) {
        for (var i in self.namespaces) {
          if (self.namespaces.hasOwnProperty(i) && '' !== i) {
              self.namespaces[i].packet({ type: 'connect' });
          }
        }
        self.publish('reconnect', self.transport.name, self.reconnectionAttempts);
      }

      clearTimeout(self.reconnectionTimer);

      self.removeListener('connect_failed', maybeReconnect);
      self.removeListener('connect', maybeReconnect);

      self.reconnecting = false;

      delete self.reconnectionAttempts;
      delete self.reconnectionDelay;
      delete self.reconnectionTimer;
      delete self.redoTransports;

      self.options['try multiple transports'] = tryMultiple;
    };

    function maybeReconnect () {
      if (!self.reconnecting) {
        return;
      }

      if (self.connected) {
        return reset();
      };

      if (self.connecting && self.reconnecting) {
        return self.reconnectionTimer = setTimeout(maybeReconnect, 1000);
      }

      if (self.reconnectionAttempts++ >= maxAttempts) {
        if (!self.redoTransports) {
          self.on('connect_failed', maybeReconnect);
          self.options['try multiple transports'] = true;
          self.transports = self.origTransports;
          self.transport = self.getTransport();
          self.redoTransports = true;
          self.connect();
        } else {
          self.publish('reconnect_failed');
          reset();
        }
      } else {
        if (self.reconnectionDelay < limit) {
          self.reconnectionDelay *= 2; // exponential back off
        }

        self.connect();
        self.publish('reconnecting', self.reconnectionDelay, self.reconnectionAttempts);
        self.reconnectionTimer = setTimeout(maybeReconnect, self.reconnectionDelay);
      }
    };

    this.options['try multiple transports'] = false;
    this.reconnectionTimer = setTimeout(maybeReconnect, this.reconnectionDelay);

    this.on('connect', maybeReconnect);
  };

})(
    'undefined' != typeof io ? io : module.exports
  , 'undefined' != typeof io ? io : module.parent.exports
  , this
);
